### PR TITLE
Changes for Version 3.4

### PR DESCRIPTION
**Pull request by NeilWJames**
These changes address the issues #256, #258, #259, #260, #261, #262, #263, #264

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -1,5 +1,19 @@
 ## Changelog
 
+### 3.4.0
+
+* SECURITY: WordPress can create images for PDF documents which if used would leak the hidden document name so image name changed.
```
Comment (benbalter) [inline docs/changelog.md:5]: Are there any remediation steps necessary for users who previously created an image for a PDF document?

File: docs/changelog.md [hunk 2]
```diff
@@ -1,5 +1,19 @@
...
+* NEW: A filter 'document_show_in_rest' is provided to display document data via the REST interface using document permissions. {#258, #259)
+* NEW: A tool is provided to validate the internal structure of all documents that the user can edit. If fixable then a button is displayed to fix it. (#260)
```
Comment (benbalter) [inline docs/changelog.md:7]: :rocket:

File: docs/shortcodes.md [hunk 3]
```diff
@@ -6,6 +6,8 @@ Existing shortcodes can be converted to and from their block forms.
 
 They are held in a grouping called `WP Document Revisions`.
 
+Since the blocks make use of dynamically-generated cont, their contents are identical whether using a block or a shortcode/widget.
```
Comment (benbalter) [inline docs/shortcodes.md:9]:
```suggestion
Since the blocks make use of dynamically-generated content, their contents are identical whether using a block or a shortcode/widget. 
```

File: includes/class-wp-document-revisions-admin.php [hunk 4]
```diff
@@ -1462,45 +1651,42 @@ public function delete_attachments_with_document( $post_id ) {
...
-					$meta = get_post_meta( $attach->ID, '_wp_attachment_metadata', true );
-
-					$file = get_attached_file( $attach->ID );
-					wp_delete_attachment( $attach->ID, false );
+		$attach = $this->get_document( $post_id );
+		if ( $attach ) {
+			// make sure that the attachment is not refered to by another post (ignore autosave).
+			global $wpdb;
+			// phpcs:disable WordPress.DB.DirectDatabaseQuery
+			$doc_link = $wpdb->get_var(
```
Comment (benbalter) [inline includes/class-wp-document-revisions-admin.php:1659]: Curious, why does this have to be done via a direct DB query?

File: includes/class-wp-document-revisions-recently-revised-widget.php [hunk 5]
```diff
@@ -68,27 +53,63 @@ public function widget_gen( $args, $instance ) {
...
+						$attach = $wpdr->get_document( $document->ID );
+						if ( $attach instanceof WP_Post ) {
+							// ensure document slug hidden from attachment.
+							$wpdr->hide_exist_doc_attach_slug( $attach->ID );
+							$image = wp_get_attachment_image( $attach->ID, 'medium' ) . '<br />';
+							if ( $std_dir !== $doc_dir ) {
+								$image = str_replace( $std_dir, $doc_dir, $image );
+							}
+						} else {
+							$image = '<p>No attachment found.</p>';
```
Comment (benbalter) [inline includes/class-wp-document-revisions-recently-revised-widget.php:103]: Should this string be translatable?

File: includes/class-wp-document-revisions-recently-revised-widget.php [hunk 6]
```diff
@@ -68,27 +53,63 @@ public function widget_gen( $args, $instance ) {
...
-				<a href="<?php echo esc_attr( $link ); ?>"><?php echo esc_html( get_the_title( $document->ID ) ); ?></a><br />
-				<?php printf( esc_html( $format_string ), esc_html( human_time_diff( strtotime( $document->post_modified_gmt ) ) ), esc_html( get_the_author_meta( 'display_name', $document->post_author ) ) ); ?>
+				<h3 class="wp-block-post-title"><a href="<?php echo esc_attr( $link ) . '"' . esc_attr( $target ) . '>' . esc_html( get_the_title( $document->ID ) ); ?></a></h3>
+				<?php
+				if ( (bool) $instance['show_thumb'] ) {
+					$thumb = get_post_thumbnail_id( $document->ID );
+					if ( $thumb ) {
+						$thumb_image     = wp_get_attachment_image_src( $thumb, 'medium' );
+						$thumb_image_alt = get_post_meta( $thumb, '_wp_attachment_image_alt', true );
+						// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
Comment (benbalter) [inline includes/class-wp-document-revisions-recently-revised-widget.php:91]: Curious why `thumb_image` and `thumb_image_alt` can't/shouldn't be escaped in this context, if it's going to be an HTML attribute? If `thumb_image_alt` is user controlled, that might allow for XSS.

File: includes/class-wp-document-revisions-validate-structure.php [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+			if ( false !== self::$parent->extract_document_id( $content ) || get_post_field( 'post_parent', $parm, 'db' ) !== $id ) {
+				return new WP_Error( 'inconsistent_parms', __( 'Inconsistent data sent to Interface', 'wp-document-revisions' ) );
+			}
+			if ( empty( $content ) || false === strpos( $content, '<' ) ) {
+				$content = (string) $parm;
+			} else {
+				$content = self::$parent->format_doc_id( $parm );
+			}
+			global $wpdb;
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.PreparedSQL.NotPrepared,WordPress.DB.DirectDatabaseQuery
```
Comment (benbalter) [inline includes/class-wp-document-revisions-validate-structure.php:133]: It might be good to include a comment indicating why this needs to be a direct DB query.

File: includes/class-wp-document-revisions-validate-structure.php [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+								unlink( $orig_dir . $sizeinfo['file'] );
+							}
+						}
+					}
+				}
+			}
+			// phpcs:enable WordPress.PHP.NoSilencedErrors.Discouraged
+		}
+
+		if ( 8 === $params['code'] ) {
```
Comment (benbalter) [inline includes/class-wp-document-revisions-validate-structure.php:224]: It may be good to document the codes and their description someplace to have them all in one place outside the `if` statement's comments.

File: includes/class-wp-document-revisions-validate-structure.php [hunk 9]
```diff
@@ -0,0 +1,628 @@
...
+							}
+						} else {
+							$ref = $beg . $ref;
+						}
+						if ( (bool) $failure['fix'] ) {
+							$fix = '<button onclick="wpdr_valid_fix(' . $failure['ID'] . ',' . $failure['code'] . ',' . $failure['parm'] . ')">' . esc_html__( 'Fix', 'wp-document-revisions' ) . '</button>';
+						} else {
+							$fix = '';
+						}
+						// phpcs:disable  WordPress.Security.EscapeOutput
```
Comment (benbalter) [inline includes/class-wp-document-revisions-validate-structure.php:344]: To avoid the security warning, could this be rewritten with inline HTML between the conditionals + string interpolation, instead of building the HTML string in PHP and outputting it all at once?

File: includes/class-wp-document-revisions.php [hunk 10]
```diff
@@ -1739,6 +1792,130 @@ public function rewrite_file_url( $file ) {
...
+	 *
+	 * @param int $attachment_id Current attachment ID.
+	 */
+	public function hide_exist_doc_attach_slug( $attachment_id ) {
+		// check that for a document.
+		$attach = get_post( $attachment_id );
+		if ( $attach->post_title !== $attach->post_name || 'document' !== get_post_type( $attach->post_parent ) ) {
+			return;
+		}
+			// has it been converted.
```
Comment (benbalter) [inline includes/class-wp-document-revisions.php:1867]: Nit: Indentation on the comment.

File: js/wp-document-revisions.coffee2 [hunk 11]
```diff
@@ -20,6 +19,8 @@
       @cookieFalse = bind(@cookieFalse, this)
       @cookieTrue = bind(@cookieTrue, this)
       @cookieDelete = bind(@cookieDelete, this)
+      @buildContent = bind(@buildContent, this)
```
Comment (benbalter) [inline js/wp-document-revisions.coffee2:22]: It's probably time to convert this to TypeScript, but that's for a different PR 😄

**Review by benbalter — APPROVED** [review]
This is really, really, great - thank you!

I left a few minor comments/questions in line. Two suggestions overall:

1. Consider adding descriptive comments any time we disable the linter (especially around direct database queries) briefly explaining why it's necessary (optimization, not possible otherwise, etc.). I suspect whomever reads this a year from now may not understand why X was necessary, and may try to "fix" it.
2. It's hard to tell from the diff, but for any of the newly added classes, consider if you think they should have, even lightweight tests, to prevent regressions. I'm not sure if that's possible, for things like the API, so I defer to your expertise.

Overall, really great, and I am excited to see this land in the next version (and can update the translation files, before it does).

File: docs/shortcodes.md (hunk 3 above)
Comment (NeilWJames) [reply docs/shortcodes.md:9]: Thanks for that.  I'll reword it to

Since the blocks make use of dynamically-generated content, the same code is used to create this as for a shortcode/widget.

since the point is that the code used is the same.

File: docs/changelog.md (hunk 1 above)
Comment (NeilWJames) [reply docs/changelog.md:5]: If the image is to be used within a shortcode/block, then before doing so, it will check if the name of the image has been changed. If it has, then it can be used. If not, then the image names are changed and then it is displayed.
(Routine hide_doc_attach_slug used to change them on loading; hide_exist_doc_attach_slug used later.)
The REST interface is more simplistic since it is not going to display the image. If it has not been changed, then the image names are hidden.

File: includes/class-wp-document-revisions-recently-revised-widget.php (hunk 6 above)
Comment (NeilWJames) [reply includes/class-wp-document-revisions-recently-revised-widget.php:91]: Thanks. Should and will be.

File: includes/class-wp-document-revisions.php (hunk 10 above)
Comment (NeilWJames) [reply includes/class-wp-document-revisions.php:1867]: Agreed

File: includes/class-wp-document-revisions-recently-revised-widget.php (hunk 5 above)
Comment (NeilWJames) [reply includes/class-wp-document-revisions-recently-revised-widget.php:103]: Yes.

File: includes/class-wp-document-revisions-validate-structure.php (hunk 9 above)
Comment (NeilWJames) [reply includes/class-wp-document-revisions-validate-structure.php:344]: I think that this is always a trade-off. I don't intrinsically think that there is a problem to put a warning exclusion (even though it allowed me to let a potential error through). 

For me the problem here was that with so many bits and pieces going into making up the URL if it went into the row definition, it woulf be difficult to see the pattern. Also looking at the variable $line (which is the document ID, ie. an internal number). It is used 5 times. So wrapping it 5 times in esc_attr(  ) seems excessive.

I have added some comments to what the cell contents are.

File: includes/class-wp-document-revisions-validate-structure.php (hunk 7 above)
Comment (NeilWJames) [reply includes/class-wp-document-revisions-validate-structure.php:133]: Two parts to this.
I started using WordPress after I retired. I worked for many years on a commercial ERP with some 2000 DB tables. SQL holds no fear for me - and I do find the continuing warnings here somewhat wearing.

However in this specific module, it is deliberate.
a) At root, my objective is to validate the document structure - which are held in the database. So I go to the database.
b) I want only 4 fields from the document - so I get just them. Yes,, I could do a WP_Query for that - but why?
c) Yes, I did hesitate over the level of access. If I gave it to admin, then clearly I would expect the `edit_post` capability would not filter out many. However, I reflected that whilst there is an amount of fixing that can be done, there may be many more times that the resolution will be to reload the document. And who should have that - those that can edit it, not administrators. Yes, this may mean that many records returned get filtered out.
d) Finally to fix the data, for metadata I use the standard functions. But for the Post, I use a direct update. Why? Because to use the WP functions would create a revision  that would create an invalid revision. This makes no sense.

I believe that a future enhancement could be to have a validate option on the "All Documents" screen. The issue for me is the management of the feedback row.

I will put some design notes at the top of the module.

File: includes/class-wp-document-revisions-validate-structure.php (hunk 8 above)
Comment (NeilWJames) [reply includes/class-wp-document-revisions-validate-structure.php:224]: OK

File: includes/class-wp-document-revisions-admin.php (hunk 4 above)
Comment (NeilWJames) [reply includes/class-wp-document-revisions-admin.php:1659]: There are 3 parts to the where condition:
a)  %d IN (post_parent, id)
You can ask for those with post_parent = X or id = X - but not in one query
b) (post_content = %d OR post_content LIKE %s )
Can't put post_content as a WP_Query parameter 
c) post_name != %s
No negative test on name field. In fact, this is actually a status selection.
It is possible that a WP_Query would not retrieve auto-save status posts, but needs to be tested.

For me, to replace this by WP_Query would be:
a) get_post(document)
if condition met, then a match. (Can bypass b))
b) WP_Query on revisions (post_parent)
loop though each testing for condition.

So it is possible. But it is a load of code.

**Comment by NeilWJames** [comment]
Thank you for your comments - and particularly for seeing the missed attributing escaping.

Regarding your comments on tests. You are absolutely correct that the code coversage has lowered significantly - and this code needs tests. I have found that the testing process somewhat exhausting seeming to go round in circles several times.

So my intention has been to first get the code (which I have tested as far as I know!) out there and then to run the gantlet of the testing system.
